Add encoder op

LiuChiachi opened the pull request:
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
OP | APIs | Models
### Description
<!-- Describe what this PR does -->
1.Encapsulated FasterTransformer's encoder inference as Paddle custom operator
2. Compile command (Optional, also support JIT)
```
cmake .. -DSM=70 -DCMAKE_BUILD_TYPE=Release -DPY_CMD=python3.6 && make -j$(nproc)
```
3.Add predict using FT encoder OP for semantic indexing example.

4.TODO
- [ ] remove padding&&rebuild padding
- [ ] FP16
- [ ] int 8

LiuChiachi committed:
add encoder op

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/PaddleNLP?pullRequest=952) <br/>All committers have signed the CLA.

FrostML commented on paddlenlp/ops/faster_transformer/sample/config/encoder.sample.yaml:
```diff
@@ -0,0 +1,46 @@
+# The parameters for topk sampling. 
+# The parameters for topp sampling. 
+# The number of decoded sentences to output.
+n_best: 1
```
`n_best` 没有用到的话，可以删除。包括上面的注释。

LiuChiachi replied:
已经删除，感谢～

FrostML commented on paddlenlp/ops/faster_transformer/sample/config/encoder.sample.yaml:
```diff
@@ -0,0 +1,46 @@
+# Number of sub-layers to be stacked in the encoder and decoder.
+n_layer: 6
+# Dropout rates.
+dropout: 0.0
```
虽然 `upscale_in_train` `.eval()` 之后都一样，如果这里要写的话，`dropout` 还是打开吧。

FrostML commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,152 @@
+        remove_padding=args.remove_padding,
+        encoder_lib=args.encoder_lib,
+        use_fp16_encoder=False)
+    encoder.load('base_trained_models/step_final/transformer.pdparams')
```
提供的 sample 这里要不还是先暂时用随机初始化的数据吧，这个 ckpt 是 `pre-norm` 的，目前 `FasterEncoder` 应该还只支持 `post-norm` 吧

guoshengCS replied:
在BERT上验证下吧，FT中这部分本来也是对BERT的加速优化

LiuChiachi replied:
感谢建议～现在已经改成了语音索引的例子。在ERNIE上进行了验证。

FrostML commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,152 @@
+    np.testing.assert_allclose(out1, out2)
+
+    with paddle.no_grad():
+        for i in range(1):
```
如果这里是想要测试性能的话，建议把 `range()` 设置大一些，多次循环求均值。

FrostML commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,152 @@
+        for i in range(1):
+            # For warmup. 
+            if 0 == i:
+                start = time.time()
```
可以再加上一个 warm up 的步数。也注意后面时间计算的时候除以的步数。

FrostML commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,152 @@
+    yaml_file = ARGS.config
+    with open(yaml_file, 'rt') as f:
+        args = AttrDict(yaml.safe_load(f))
+        pprint(args)
```
`pprint` 可以放在 args 都处理完后，`args.use_fp16_encoder` 后面。

LiuChiachi replied:
感谢建议，已经修改～

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+#include <vector>
+
+
+#include <iostream>
```
可以确认下 iostream 有否用到

LiuChiachi replied:
感谢建议，已经删除～

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+  if (input.place() == paddle::PlaceType::kGPU) {
+    auto shape = input.shape();
+
+    std::vector<int64_t> output_dims = {shape[0], shape[1], shape[2]};
```
这里可以直接使用 `input.shape()` 吧

LiuChiachi replied:
是的，感谢建议。已经修改～

FrostML replied:
似乎没有修改？看看是不是改漏了

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+    const bool& allow_gemm_test,
+    const bool& use_trt_kernel,
+    const int64_t& max_seq_len) {
+  return {{input_shape[0], input_shape[1], input_shape[2]}};
```
同上，可以直接用 `input_shape` 作为返回哈

LiuChiachi replied:
感谢建议，已经修改

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+    const paddle::DataType& sequence_id_offset,
+    const paddle::DataType& trt_seqlen_offset,
+    const paddle::DataType& amax_list) {
+  switch (input) {
```
同上，也直接使用 `input` 作为 dtype 返回

LiuChiachi replied:
感谢，已经修改

FrostML replied:
同上，似乎没有修改？看看是不是改漏了

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -0,0 +1,289 @@
+  encoder->initialize(encoder_param);
+  encoder->forward();
+  encoder->freeBuffer();
+  delete allocator_;
```
加上 `delete encoder;` 吧~

LiuChiachi replied:
感谢，已经修改～

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,337 @@
+            layer_weight.append(mod.linear2.weight)
+            layer_weight.append(mod.linear2.bias)
+            # if not self.int8_mode:
+            layer_weight.append(paddle.to_tensor([]))
```
如果是想传一个空的 tensor 进去，可以用 `paddle.zeros([])`，`to_tensor` 静态图应该用不了。

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,337 @@
+                axis=0,
+                dtype="int32")
+        else:
+            sequence_id_offset = paddle.to_tensor([], dtype="int32")  # 不需要
```
空的 tensor 可以用 `paddle.zeros([])`

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,337 @@
+            dtype = "float16"
+            enc_input = paddle.cast(enc_input, dtype)
+            src_mask = paddle.cast(src_mask, dtype)
+            mem_seq_lens = paddle.cast(mem_seq_lens, dtype)
```
`mem_seq_lens` 不需要 cast 吧

FrostML commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,152 @@
+    batch_size = args.batch_size
+
+    src_word = paddle.randint(
+        0, 30000, shape=[args.batch_size, args.seq_len], dtype='int64')
```
30000 可以直接使用 `args.src_vocab_size`

Commit:
update

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,337 @@
+                 remove_padding=False,
+                 encoder_lib=None,
+                 use_fp16_encoder=False):
+        super().__init__()
```
`super(FasterEncoder, self).__init__()`?

LiuChiachi replied:
Done. Thanks!

Commit:
support fp16

Commit:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-tmp

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,333 @@
+            use_fp16_encoder=use_fp16_encoder)
+
+    def forward(self, enc_input, src_mask, mem_seq_lens):
+        self.encoder.eval()
```
不要在 `forward` 里面 `.eval`，`eval` 了再进 `forward`.

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,333 @@
+        enc_out = self.encoder(enc_input, src_mask, mem_seq_lens)
+
+        if self.use_fp16_encoder:
+            enc_out = paddle.cast(enc_out, "float32")
```
如果 `enc_out` 后面要输入 fp16 的 decoding 或是 decoder，建议可以考虑把这里的 cast 设置成可选的，避免重复 cast。

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,333 @@
+            dropout=dropout,
+            custom_encoder=custom_encoder,
+            activation="gelu",
+            normalize_before=False)
```
这里的 `self.transformer` 是否是有些多余？如果需要这个 encoder 的实例，可以把 `custom_encoder` 直接拿给 `InferTransformerEncoder(encoder=...)` 或是直接接受外部传入的 encoder。

LiuChiachi replied:
感谢建议，现在已经修改了这一部分。

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,333 @@
+        return enc_out
+
+
+class FasterEncoder(nn.Layer):
```
`FasterEncoder` 这个 API 和上面的 `InferTransformerEncoder` 似乎是没有区别？考虑是否把这里封装成针对 model 的接口，上面的 `InferTransformerEncoder` 作为 encoder 的接口。

LiuChiachi replied:
感谢建议，现在已经修改为语义索引的例子

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+        "TRTSeqLenOffset",
+        "AmaxList",
+    })
+    .Outputs({"encoder_out"})
```
输入输出命名风格最好统一下，和上面Input一致

LiuChiachi replied:
Done. Thanks

guoshengCS commented on paddlenlp/ops/faster_transformer/sample/encoder_sample.py:
```diff
@@ -0,0 +1,154 @@
+            start=0, end=src_max_len)
+
+    src_emb = encoder.src_word_embedding(src_word)
+    src_pos_emb = encoder.src_pos_embedding(src_pos)
```
embedding这些感觉不用在接口里封装，而且封装了这里确实也没有在encoder内部使用而是拿出来用的，这个其实也和我们的一贯用法不太相符。对于encoder或者encoder layer我们更期望的接口形态是不新增接口，原有TransformerEncoder/TransformerEncoderLayer在满足条件时能直接没有任何新增使用成本的切换到快速预测版本。

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cc:
```diff
@@ -0,0 +1,202 @@
+            "layer_idx: int64_t",
+            "allow_gemm_test: bool",
+            "use_trt_kernel: bool",
+            "max_seq_len: int64_t"})
```
这里是必须要加`max_seq_len`这样一个属性吗，看FT里length是使用的mask输入的最后两维

guoshengCS commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,333 @@
+
+        self.src_word_embedding = WordEmbedding(
+            vocab_size=src_vocab_size, emb_dim=d_model, bos_id=self.bos_id)
+        self.src_pos_embedding = PositionalEmbedding(
```
forward中没有使用这些embedding的话就不要在这里包括了

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -0,0 +1,289 @@
+      reinterpret_cast<const DataType_*>(output_layernorm_bias.data<data_t_>());
+  int valid_word_num;
+  if (remove_padding == true) {
+    valid_word_num = output_layernorm_weight.shape()[0];
```
这个weight的shape是固定大小应该不能作为valid_word_num吧，虽然还没支持也尽量注意

Commit:
update encoder sample

remove useless config

use pad to max seq len

FrostML commented on paddlenlp/ops/faster_transformer/src/CMakeLists.txt:
```diff
@@ -270,7 +270,11 @@ else(ON_INFER)
 
+  add_library(encoder_op SHARED ${encoder_op_files})
+  add_dependencies(encoder_op extern_${THIRD_PARTY_NAME} boost)
+  target_link_libraries(encoder_op PRIVATE -lcublas -lcudart ${lib_link} ${ft_lib_link} -lencoder -lencoder_gemm_func -lencoder_igemm_func -ltrt_fused_multi_head_attention)
```
可以确认下这里的 lib 是否都有用到，`${ft_lib_link}` 里面还包含有 decoding 相关的 lib，另外 `encoder_gemm_func` 和 `encoder_igemm_func` 似乎是 FT 单测的 so

FrostML commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -0,0 +1,290 @@
+#include <curand.h>
+#include <curand_kernel.h>
+#include <algorithm>
+#include <iostream>
```
确认下 `#include <iostream>` 是否有用到？

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,239 @@
+            layer_weight.append(mod.linear2.weight)
+            layer_weight.append(mod.linear2.bias)
+            # if not self.int8_mode:
+            layer_weight.append(paddle.to_tensor([]))
```
`paddle.to_tensor([])` 静态图用不了，@guoshengCS 看看这个 API 本身是否需要动静统一

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,239 @@
+                axis=0,
+                dtype="int32")
+        else:
+            sequence_id_offset = paddle.to_tensor([], dtype="int32")
```
同上，`paddle.to_tensor([])` 静态图用不了，@guoshengCS 看看这个 API 本身是否需要动静统一

FrostML commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,239 @@
+            trt_seq_len = paddle.reshape(t, [-1])
+            trt_seq_len = paddle.concat(
+                [
+                    trt_seq_len, paddle.to_tensor(
```
同上，`paddle.to_tensor([])` 静态图用不了，@guoshengCS 看看这个 API 本身是否需要动静统一

guoshengCS commented on paddlenlp/ops/faster_transformer/src/fusion_encoder_op.cu:
```diff
@@ -0,0 +1,290 @@
+    encoder_param.amaxList = nullptr;
+  }
+  BertEncoderTransformer<EncoderTraits_>* encoder =
+      new BertEncoderTransformer<EncoderTraits_>(int8_mode, allow_gemm_test);
```
注意确认下这里的耗时，NV FT中FT和Torch的OP将这个构造都放在了forward外面 https://github.com/NVIDIA/FasterTransformer/blob/v3.1/fastertransformer/tf_op/bert_transformer_op.cc#L84 、 https://github.com/NVIDIA/FasterTransformer/blob/v3.1/fastertransformer/th_op/encoder_ext.h#L83 ，应该是call once的，在forward时另外通过拷贝构造了一个 https://github.com/NVIDIA/FasterTransformer/blob/v3.1/fastertransformer/tf_op/bert_transformer_op.cc#L180 ，而且按照NV FT4.0的注释 https://github.com/NVIDIA/FasterTransformer/blob/v4.0/fastertransformer/tf_op/bert_transformer_op.cc#L103 这里会有一定的耗时，看里面有读取配置文件外和getSMVersion这种调用。如果确实是这样的话目前咱们不能类似TF、Torch这样写，可能就得优化BertEncoderTransformer构造函数了。

LiuChiachi committed:
commit before pull

LiuChiachi committed:
solve conflicits

LiuChiachi committed:
update patches and cmakelist

LiuChiachi committed:
upadte the interface for encoder op

LiuChiachi committed:
support jit, and support replacing forward function of nn.TransformerEncoder(and *Layer) object

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-tmp

LiuChiachi committed:
support jit

move encoder sample to semantic index, and update readme

delete fp16

add device config info

fix disable faster encoder bug

guoshengCS commented on examples/semantic_indexing/README.md:
```diff
@@ -232,6 +232,33 @@ python -u -m paddle.distributed.launch --gpus "0" \
+
+执行上述操作后，可以得到与原生预测非常接近的结果，在float32下，且batch_size=32, max_seq_len=64时，二种方式预测下最终余弦相似度的最大绝对误差约为3.93e-6。
+
+通过比较，可以得到在不同batch_size, max_seq_len下，使用集成了FasterTransformer的高性能算子可以对Encoder部分的推理进行加速（其余参数都与默认值相同）。测得部分性能数据如下，从表中可以看出在更小的batch_size和max_seq_len上，使用FasterTransformer预测更有优势。
```
说明下使用的机器环境

guoshengCS commented on examples/semantic_indexing/faster_predict.py:
```diff
@@ -0,0 +1,277 @@
+    def load(self, init_from_params):
+        if self.use_fp16_encoder:
+            for mod in self.ptm.encoder.layers:
+                mod.self_attn.q_proj.weight = transfer_param(
```
这个load最好是作为encoder/encoder_layer接口提供，用户sample不要有这么复杂的代码。

guoshengCS commented on examples/semantic_indexing/README.md:
```diff
@@ -232,6 +232,33 @@ python -u -m paddle.distributed.launch --gpus "0" \
+| 16         | 64          | 33.31318140029907   | 52.44770574569702   |
+| 32         | 16          | 12.891342163085938  | 22.621662139892578  |
+| 32         | 32          | 17.206310987472534  | 22.18772006034851   |
+| 32         | 64          | 28.032865524291992  | 24.947681188583374  |
```
这个变慢的setting性能还要关注确认下

guoshengCS commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,194 @@
+def disable_faster_encoder(self):
+    def init_func(layer):
+        if isinstance(layer, (TransformerEncoderLayer, TransformerEncoder)):
+            layer.forward = type(layer).forward
```
用class的forward方法成员给该class对象的forward赋值会有些问题，要用对象的成员来赋值

LiuChiachi replied:
感谢提醒，已经修改:)

guoshengCS commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,194 @@
+                          src_mask,
+                          cache=None,
+                          sequence_id_offset=None,
+                          mem_seq_lens=None,
```
这个mem_seq_lens在这里没有使用到是吗

LiuChiachi replied:
是的，已经删去了，感谢提醒

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-tmp

guoshengCS commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,194 @@
+            trt_seq_len, paddle.to_tensor(
+                [batch_size * max_seq_len], dtype=trt_seq_len.dtype)
+        ],
+        axis=0)
```
这些目前不支持但还是带上的操作看着有点太多了，看FT encoder fp32时应该不依赖这些具体的值，看看是否能简化下吧

LiuChiachi replied:
感谢，已经删除

LiuChiachi committed:
remove unsupported argument

add docstring and warning process for False bias_attr of EncoderLayer

fix docstring format

add \n, and add docstring for infer_transformer_encoder

guoshengCS commented on paddlenlp/ops/faster_transformer/transformer/encoder.py:
```diff
@@ -0,0 +1,162 @@
+
+
+def enable_faster_encoder(self):
+    def init_func(layer):
```
稍微补充些文档吧，

guoshengCS replied:
另外`enable_faster_encoder`最好做下匹配处理，在FT encoder无法支持（如normlaize_before、multiheadattention中linear有bias）以及jit编译失败还是用原来的forward

LiuChiachi replied:
已经补充了对linear中bias_attr为False，以及对jit编译失败的处理。`normlaize_before`支持的PR会在本PR合入后迅速支持，这次先省去了对`normlaize_before`的异常处理。

guoshengCS commented on examples/semantic_indexing/faster_predict.py:
```diff
@@ -0,0 +1,230 @@
+            self.emb_reduce_linear = paddle.nn.Linear(
+                768, output_emb_size, weight_attr=weight_attr)
+        try:
+            load("FasterTransformer", verbose=True)
```
这个jit编译放到`enable_faster_encoder`下面吧，并且在`enable_faster_encoder`里面做下编译失败的切换处理

LiuChiachi replied:
感谢提醒，已经修改～
